gateway/ln-gateway/src/bin/cln_extension.rs

```diff
@@ -0,0 +1,330 @@
+                exclude: None,
+                maxfee: None,
+                description: None,
+            }))
+            .await
+            .map(|response| {
+                let pay_success = match response {
+                    cln_rpc::Response::Pay(model::PayResponse {
+                        payment_preimage, ..
+                    }) => PayInvoiceResponse {
+                        preimage: payment_preimage.to_vec(),
+                    },
+                    _ => panic!("Unexpected response from cln_rpc"),
+                };
+                tonic::Response::new(pay_success)
+            })
+            .map_err(|e| {
+                error!("cln pay returned error {:?}", e);
+                Status::internal(e.to_string())
+            })
+    }
+
+    type SubscribeInterceptHtlcsStream =
+        ReceiverStream<Result<SubscribeInterceptHtlcsResponse, Status>>;
+
+    async fn subscribe_intercept_htlcs(
+        &self,
+        request: tonic::Request<SubscribeInterceptHtlcsRequest>,
+    ) -> Result<tonic::Response<Self::SubscribeInterceptHtlcsStream>, Status> {
+        let SubscribeInterceptHtlcsRequest { short_channel_id } = request.into_inner();
+        let receiver = self
+            .interceptor
+            .lock()
+            .await
+            .add_htlc_subscriber(short_channel_id)
+            .await;
+
+        Ok(tonic::Response::new(ReceiverStream::new(receiver)))
+    }
+}
+
+#[derive(Debug)]
+pub struct LightningError(pub Option<i32>);
+
+#[derive(Debug, Error)]
+pub enum ClnRpcError {
+    #[error("ConfigurationError")]
+    ConfigurationError,
+    #[error("RpcServerError : {0:?}")]
+    RpcServerError(LightningError),
+    #[error("Other: {0:?}")]
+    Other(#[from] anyhow::Error),
+}
+
+/// Functional structure to filter intercepted HTLCs into subscription streams.
+/// Used as a CLN plugin
+struct ClnHtlcInterceptor {
+    subscriptions:
+        Mutex<HashMap<u64, mpsc::Sender<Result<SubscribeInterceptHtlcsResponse, Status>>>>,
+}
+
+impl ClnHtlcInterceptor {
+    fn new() -> Self {
+        Self {
+            subscriptions: Mutex::new(HashMap::new()),
+        }
+    }
+
+    async fn intercept_htlc(&self, payload: HtlcAccepted) -> serde_json::Value {
+        if let Some(subscription) = self
+            .subscriptions
+            .lock()
+            .await
+            .get(&payload.onion.short_channel_id)
+        {
+            match subscription
+                .send(Ok(SubscribeInterceptHtlcsResponse {
+                    payment_hash: payload.htlc.payment_hash.to_vec(),
+                    amount: payload.htlc.amount.msats,
+                    units: "msat".to_string(),
+                    expiry: payload.htlc.cltv_expiry,
+                }))
+                .await
+            {
+                Ok(_) => {
+                    // TODO: Return outcome of HTLC processing by the gateway
+                    return json!({ "result": "continue" });
```

Comment:
Shouldn't the contents of `Ok` be the preimage? Then just put that inside the JSON?

Reply:
WIP. Working on the todo highlighted to return `Ok` JSON with preimage